firmware/library/L1_Drivers/can.hpp

```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#include "L0_LowLevel/LPC40xx.h"
+#include "L2_Utilities/macros.hpp"
+
+SJ2_PACKED(struct) CanMessage_t
+{
+    union
+    {
+        uint32_t frame;
+        struct
+        {
+            uint8_t tx_priority       : 8;
+            uint8_t                   : 8;      // reserved
+            uint8_t data_length       : 4;
+            uint16_t                  : 10;     // reserved
+            uint8_t remote_tx_request : 1;
+            uint8_t frame_format      : 1;
+        } frame_data;
+    };
+    uint32_t id;
+    uint8_t data[8];
+};
+
+class CanInterface
+{
+ public:
+    enum TestType
+    {
+        kLocal,
+        kGlobal
+    };
+
+    virtual bool Initialize()                             = 0;
+    virtual bool Send(const CanMessage_t &message)        = 0;
+    virtual bool Receive(CanMessage_t * message)          = 0;
+    virtual bool SelfTestBus(const TestType kType,
+                             const CanMessage_t &message) = 0;
```

Comment:
Do you need a CanMessage to do a self test?

Reply:
Yes. I could also make it so that it's hardcoded to some default message internal to the driver and then all the user has to do is call `SelfTestBus(local)`. I think that would be preferred since the whole point of self-test is to verify everything checks out on the hardware level.

Reply:
removed message paramter